tests/e2e/tests/stackdriver/stackdriver_test.go

```diff
@@ -0,0 +1,282 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stackdriver
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	"cloud.google.com/go/logging/logadmin"
+	"fortio.org/fortio/fhttp"
+	"fortio.org/fortio/periodic"
+	"golang.org/x/oauth2/google"
+	cloudtrace "google.golang.org/api/cloudtrace/v1"
+	"google.golang.org/api/iterator"
+	monitoring "google.golang.org/api/monitoring/v3"
+
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/tests/e2e/framework"
+	"istio.io/istio/tests/util"
+)
+
+const (
+	fortioYaml  = "tests/e2e/tests/stackdriver/fortio-rules.yaml"
+	adapterYaml = "tests/e2e/tests/stackdriver/adapter.yaml"
+)
+
+type (
+	fortioTemplate struct {
+		FortioImage string
+	}
+	adapterTemplate struct {
+		Namespace string
+	}
+	testConfig struct {
+		*framework.CommonConfig
+	}
+)
+
+var (
+	tc      *testConfig
+	gcpProj *string
+)
+
+func (t *testConfig) Setup() error {
+	if !util.CheckPodsRunning(t.Kube.Namespace, t.Kube.KubeConfig) {
+		return fmt.Errorf("could not get all pods running")
+	}
+
+	filled, err := util.CreateAndFill(
+		t.Kube.TmpDir,
+		util.GetResourcePath(adapterYaml),
+		&adapterTemplate{
+			Namespace: t.Kube.Namespace,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	if err := util.KubeApply(t.Kube.Namespace, filled, t.Kube.KubeConfig); err != nil {
+		log.Errorf("Kubectl apply adapter yaml failed.")
+		return err
+	}
+
+	gateway, errGw := tc.Kube.IngressGateway()
+	if errGw != nil {
+		return errGw
+	}
+	if _, err := sendTrafficToCluster(gateway); err != nil {
+		return fmt.Errorf("generating HTTP traffic failed: %v", err)
+	}
+
+	// Push interval is 5s, wait for metrics to be written
+	time.Sleep(10 * time.Second)
+
+	return nil
+}
+
+func sendTrafficToCluster(gateway string) (*fhttp.HTTPRunnerResults, error) {
+	opts := fhttp.HTTPRunnerOptions{
+		RunnerOptions: periodic.RunnerOptions{
+			QPS:        10,
+			Exactly:    60,
+			NumThreads: 5,
+			Out:        os.Stderr,
+		},
+		HTTPOptions: fhttp.HTTPOptions{
+			URL: gateway + "/?status=418:10,520:15&size=1024:10,512:5",
+		},
+		AllowInitialErrors: true,
+	}
+	return fhttp.RunHTTPTest(&opts)
+}
+
+func (t *testConfig) Teardown() error {
+	return nil
+}
+
+func TestMain(m *testing.M) {
+	gcpProj = flag.String("gcp_proj", os.Getenv("GCP_PROJ"), "GCP Project ID, required")
+	flag.Parse()
+	check(framework.InitLogging(), "cannot setup logging")
+	check(setTestConfig(), "could not create TestConfig")
+	tc.Cleanup.RegisterCleanable(tc)
+	os.Exit(tc.RunTest(m))
+}
+
+func TestContextGraph(t *testing.T) {
+	// FIXME add test when read API is available.
```

Comment:
should we add a `t.Skip("ContextGraph API testing not yet supported")` here?

Reply:
I had that, but it requires a GitHub issue. I thought it doesn't make sense to have that as it's more of a Google issue than an Istio issue.